### ISSUE TITLE
Error loading python36.dll

### ISSUE THREAD

**Issue by ElDifinitivo** [qutebrowser/qutebrowser#2788]
I'm trying to test out v0.11 on a Windows installation, but I can't get it to run.  I receive a fatal error message that states:

```
Error loading Python DLL:
D:\qutebrowser-0.11.0-windows-standalone-amd64\qutebrowser-0.11.0-x64\python36.dll (error code 193)
```

I'm trying to have it run off of my USB (as I have with v0.10 due to lacking admin rights on my work computer), so I've been using the standalone version.

Running on Windows 7 Professional w/Service Pack 1.

**Comment by The-Compiler** [comment]
I just tried in a Windows 10 VM, a Windows 7 VM and a physical Windows 7 machine, without any issues.

Some ideas:

- Try re-downloading and re-extracting it
- Try running it from outside the USB stick
- Try disabling security/anti-virus software which may prevent it from running (if you trust me, that is 👿)

Also, I'm assuming you are on a 64bit machine? Maybe [dependency walker](http://dependencywalker.com/) shows something interesting on `qutebrowser.exe` or `python36.dll`?

**Comment by SavoyRoad** [comment]
Re-downloading/extracting onto the computer itself didn't do anything, and there's no security software running on here that would cause anything strange.

Dependency walker definitely spit out some error messages, though.  For `qutebrowser.exe`, I get 'Error opening file. The system cannot find the file specified (2).` for the following modules:

```
API-MS-WIN-CORE-WINRT-ERROR-L1-1-0.DLL
API-MS-WIN-CORE-WINRT-L1-1-0.DLL
API-MS-WIN-CORE-WINRT-ROBUFFER-L1-1-0.DLL
API-MS-WIN-CORE-WINRT-STRING-L1-1-0.DLL
DCOMP.DLL
IESHIMS.DLL
```

These two additional warnings are also present:
```
Warning: At leat one delay-load dependency module was not found.
Warning: At least one module has an unresolved import due to a missing export function in a delay-load dependent module.
```

For `python36.dll`, it's even messier.  The same 'Error opening file. The system cannot find the file specified (2).` for these modules:

```
API-MS-WIN-CRT-CONIO-L1-1-0.DLL
API-MS-WIN-CRT-CONVERT-L1-1-0.DLL
API-MS-WIN-CRT-ENVIRONMENT-L1-1-0.DLL
API-MS-WIN-CRT-FILESYSTEM-L1-1-0.DLL
API-MS-WIN-CRT-HEAP-L1-1-0.DLL
API-MS-WIN-CRT-LOCALE-L1-1-0.DLL
API-MS-WIN-CRT-MATH-L1-1-0.DLL
API-MS-WIN-CRT-PROCESS-L1-1-0.DLL
API-MS-WIN-CRT-RUNTIME-L1-1-0.DLL
API-MS-WIN-CRT-STDIO-L1-1-0.DLL
API-MS-WIN-CRT-STRING-L1-1-0.DLL
API-MS-WIN-CRT-TIME-L1-1-0.DLL
API-MS-WIN-CORE-WINRT-ERROR-L1-1-0.DLL
API-MS-WIN-CORE-WINRT-L1-1-0.DLL
API-MS-WIN-CORE-WINRT-ROBUFFER-L1-1-0.DLL
API-MS-WIN-CORE-WINRT-STRING-L1-1-0.DLL
DCOMP.DLL
IESHIMS.DLL
```

as well as these additional warnings:

```
Error: At least one required implicit or forwarded dependency was not found.
Error: At least one module has an unresolved import due to a missing export function in an implicitly dependent module.
Error: Modules with different CPU types were found.
Warning: At least one delay-load dependency module was not found.
Warning: At least one module has an unresolved import due to a missing export function in a delay-load dependent module.
```

The module with a different CPU type is linked to `VCRUNTIME140.DLL`.

**Comment by The-Compiler** [comment]
The `API-MS-WIN-` ones are fine, and `DCOMP.DLL` / `IESHIMS.DLL` are probably as well.

The `VCRUNTIME140.DLL` with a different CPU type sounds worrying though. Are you sure this is a 64bit windows? Can you try whether the 32bit version of qutebrowser works?

**Comment by SavoyRoad** [comment]
It's definitely 64-bit, but I had to ask an IT admin to install `VCRUNTIME140.DLL` as it was missing when I originally had tried to run it.  I'll bother them again to install the 64-bit version and see if it fixes any issues.

**Comment by SavoyRoad** [comment]
So according to IT, both x86 and x64 are installed.  I don't have enough experience with Windows to know where to go from there though.

**Comment by The-Compiler** [comment]
Maybe qutebrowser could (should?) also ship with that DDL. Have you tried the 32bit version though?

**Comment by SavoyRoad** [comment]
Trying with the 32bit version spits out this error message:
```
The program can't start because
api-ms-win-crt-runtime-|1-1-0.dll is missing from your computer.
Try reinstalling the program to fix this problem.
```
followed by the same `python36.dll` message, this time `error code 126`

**Comment by The-Compiler** [comment]
So I did a quick search for the DLL on my machine, and found four of them. I uploaded them to https://www.qutebrowser.org/tmp/vcruntime

Can you try them one by one, renaming them to `vcruntime140.dll` and placing them into the same directory the `qutebrowser.exe` is? Make sure to try the 64bit ones with the 64bit standalone zip, and the 32bit ones with the 32bit.

If that doesn't work, can you please also try finding the `vcruntime140.dll` from your system (probably in `C:\Windows\system32` for 64bit [sic!], and `syswow64` for 32bit) and copying it into the qutebrowser directory?

**Comment by SavoyRoad** [comment]
Okay, so both 32bit `.dll` files will cause `qutebrowser.exe` to repeat the same 
```
The program can't start because
api-ms-win-crt-runtime-|1-1-0.dll is missing from your computer.
Try reinstalling the program to fix this problem.
```
from above, followed by the `python36.dll` error message.

`64bit-python_vcruntime140.dll` produces the same output as the 32bit versions, while `64bit-system_vcruntime140.dll` doesn't mention anything about the `api-ms-win-crt-runtime*.dll` but still mentions the `python36.dll` error.

The output for the local `vcruntime140.dll` for 64bit mimics that of `64bit-system_vcruntime140.dll`, while the 32bit version mimics both 32bit `.dll` files you provided.

**Comment by The-Compiler** [comment]
I'm out of ideas then, sorry. Let's see if anyone else sees the same.

**Comment by beriain** [comment]
It's not the same situation, but hope it helps.

I had also the python36.dll error message on Windows 8.1 64bit with the 64bit installer, not the standalone version.

I solved it installing the latest Python 3.6.1 64bit

**Comment by The-Compiler** [comment]
@Link-Satonaka Do you have any idea why what @beriain would happen with the installer only?

**Comment by The-Compiler** [comment]
https://github.com/takluyver/pynsist/issues/90 talks about some similar issues, and apparently they go away when a Windows install is fully updated. @SavoyRoad @beriain when did you last update? :wink:

Also, can you try to download [the pynsist repo](https://github.com/takluyver/pynsist/archive/master.zip) and copying all DLLs from `nsist/msvcrt/x64/` (or `/x86`) to the qutebrowser folder and see if that changes anything?

**Comment by SavoyRoad** [comment]
Updates are all controlled by the sysadmin, so I can't do a thing about that :P  
However, last check for updates was earlier today, and the last updates were installed on 15 May.  Not sure if there's been anything else since as all I can get is that Windows is up to date as of now.

Using all of the DLLs from the pynsist repo spits out the same `error code 193` for the `python36.dll` :/

**Comment by The-Compiler** [comment]
https://github.com/pyinstaller/pyinstaller/issues/1566#issuecomment-269741477 might be relevant oo.

**Comment by The-Compiler** [comment]
@SavoyRoad Can you try with the exe's from https://qutebrowser.org/tmp/aelius please? I recently built them for someone else, but PyInstaller has a recent fix which might be relevant here.

**Comment by SavoyRoad** [comment]
I tried out both standalone versions.  I could not unzip the 64 bit file (complained about requiring a mutli-disk archive), while the 32 bit version showed the same dll errors from earlier in the thread.

**Comment by Aeres-u99** [comment]
Tried both standalone and zip for 64bit computers
Tried reinstalling vcredist after downloading it from microsoft 
its still throwing me the very same error 
<code> The procedure entry point ucrtbase.abort could not be located in the dynamic link library api-ms-win-crt-runtime-l1-1-0.dll</code>
and after that <code> Error loading DLL D:\path\to\browser\python36.dll.LoadLibrary: The specified procedure could not be found.</code>

**Comment by bitraid** [comment]
@Aeres-u99 which version of Windows are you using?

**Comment by bitraid** [comment]
Your system is probably missing some files needed by programs that use the new c runtimes. The update [KB2999226 for Windows 7 x64](https://www.microsoft.com/en-us/download/details.aspx?id=49093) has these files, but it requires Service Pack 1. 
Most programs that use the new crt though, they do ship with these files. You can try to copy the files `api-ms-win-*.dll` (about 40 small files) from the directory of a program that already has them (firefox for example) to qutebrowser's directory.

**Comment by The-Compiler** [comment]
This should be fixed in qutebrowser v1.6.0 which will be released over the next few days.
